diff --git a/haskell/cabal.bzl b/haskell/cabal.bzl
index d8691a20..475fca37 100644
--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -93,6 +93,9 @@ def _prepare_cabal_inputs(hs, cc, dep_info, cc_info, component, package_id, tool
     with_profiling = is_profiling_enabled(hs)
 
     (ghci_extra_libs, env) = get_ghci_extra_libs(hs, cc_info)
+    env = {
+        k: ":".join(v.split(";"))
+    }
     env["PATH"] = ":".join([paths.dirname(cc.tools.ar), _make_path(hs, tool_inputs.to_list())])
     if hs.toolchain.is_darwin:
         env["SDKROOT"] = "macosx"  # See haskell/private/actions/link.bzl
